### PR TITLE
Add external access APIs for legacy code analysis

### PR DESCRIPTION
**Pull request by sharwell**
Implements the APIs required to complete #35080.

### CODE REVIEW

File: src/Tools/ExternalAccess/CodeLens/Microsoft.CodeAnalysis.ExternalAccess.CodeLens.csproj [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    <PackageId>Microsoft.CodeAnalysis.ExternalAccess.CodeLens</PackageId>
+    <PackageDescription>
+      A supporting package for CodeLens:
+      [Link needed]
+    </PackageDescription>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <!--
+      ⚠ ONLY CODELENS ASSEMBLIES MAY BE ADDED HERE ⚠
```
Comment (jasonmalinowski) [inline src/Tools/ExternalAccess/CodeLens/Microsoft.CodeAnalysis.ExternalAccess.CodeLens.csproj]: This stuff builds directly in the VS repo, doesn't it? We're also planning on doing ExternalAccess libraries for those? I thought we weren't, but would understand why you would want to. This is more a "if we're changing the plan, just want to know that we are changing the plan" question.

File: src/Tools/ExternalAccess/LegacyCodeAnalysis/LegacyCodeAnalysisVisualStudioDiagnosticAnalyzerService.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.LanguageServices.Implementation.Diagnostics;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.LegacyCodeAnalysis
+{
+    [Export(typeof(ILegacyCodeAnalysisVisualStudioDiagnosticAnalyzerService))]
```
Comment (jasonmalinowski) [inline src/Tools/ExternalAccess/LegacyCodeAnalysis/LegacyCodeAnalysisVisualStudioDiagnosticAnalyzerService.cs]: Are we generally going to do the interface plus concrete type approach for all of this, or just want to directly export the concrete type? Usually for internally-sort of things where we're less interested in hiding implementation we've just done concrete types.

No harm of course doing it this way other than just more files and more code.

File: src/Tools/ExternalAccess/CodeLens/Microsoft.CodeAnalysis.ExternalAccess.CodeLens.csproj (hunk 1 above)
Comment (sharwell) [reply src/Tools/ExternalAccess/CodeLens/Microsoft.CodeAnalysis.ExternalAccess.CodeLens.csproj]: This is generalized by the build boss rules. The new approach applies to everything which:

1. Does not build in dotnet/roslyn
2. Ships with or loads within Visual Studio process(es)

These rules ensure that PR builds of Roslyn are correct even before insertion-time checks.

File: src/Tools/ExternalAccess/LegacyCodeAnalysis/LegacyCodeAnalysisVisualStudioDiagnosticAnalyzerService.cs (hunk 2 above)
Comment (sharwell) [reply src/Tools/ExternalAccess/LegacyCodeAnalysis/LegacyCodeAnalysisVisualStudioDiagnosticAnalyzerService.cs]: ➡️ Submitted #35727 to document the patterns I've started to settle on

**Comment by sharwell** [comment]
Superseded by #36151

**Pull request closed without merging**